### ISSUE TITLE
[shapes] Missing pixels on `DrawCircle()` with small radius and window `FLAG_MSAA_4X_HINT` active

### ISSUE THREAD

**Issue by evanrinehart** [raysan5/raylib#3247]
```C
#include <raylib.h>

int main(int argc, char* argv[]){

    SetConfigFlags(FLAG_MSAA_4X_HINT);
    InitWindow(800, 600, "TEST CASE");

    while(!WindowShouldClose()){
        BeginDrawing();
        ClearBackground(WHITE);
        DrawCircle(400.5, 300.5, 8, BLACK);
        DrawCircle(528.0, 172.0, 8, BLACK);
        EndDrawing();
    }

    CloseWindow();

    return 0;
}
```

![image](https://github.com/raysan5/raylib/assets/1022635/f2bbf8fb-87ab-4726-b35e-6e40671ebc50)

This is on latest Ubuntu with NVIDIA GeForce GTX 1080 Ti. It seems like the constituents of the disc don't line up exactly and MSAA_4X has found those spots not covered.

**Comment by evanrinehart** [comment]
I narrowed it down to the rshapes.c routine DrawCircleSector. Specifically the SUPPORT_QUADS_DRAW_MODE branch of that code. When I force the triangles version of DrawCircleSector to draw my discs, the visual errors go away.

Something is going wrong when trying to draw circles with quads

**Comment by raysan5** [comment]
@evanrinehart I'm aware of this issue. I don't know why it happens...

**Comment by krupitskas** [comment]
Can this be OpenGL bug?

I've briefly took a look:
When we are drawing to MSAA target samples look like this:
Sample 0:
<img width="180" alt="image" src="https://github.com/raysan5/raylib/assets/5683877/a752abbc-9db1-490f-a5ef-7b18ec9ba971">
Sample 1:
<img width="176" alt="image" src="https://github.com/raysan5/raylib/assets/5683877/c6935761-31cb-4d87-83d7-d5e793a909f8">
Other samples are having same issues, thats why resolve does look like this:
<img width="181" alt="image" src="https://github.com/raysan5/raylib/assets/5683877/2131747d-9739-4b3c-a0b4-69da0af232df">

However I've checked shader and looks like issue with sampling of provided texture in fragment shader
```
    vec4 texelColor = texture(texture0, fragTexCoord);   
    finalColor = texelColor*colDiffuse*fragColor;        
```

I see that texture0 is R8G8_Unorm with RRRG swizzling, so I've tried to replicate same sampling in the shader

```
vec4 texelColor = texture(texture0, fragTexCoord).rrrg;   
```

And its gone...

<img width="188" alt="image" src="https://github.com/raysan5/raylib/assets/5683877/eb6d150c-18d3-4810-91d8-06d1fa2f8966">

**Comment by raysan5** [comment]
@krupitskas thanks for the investigation and the detailed explanation. I'm not sure about the reasons for this issue but my guess is that could be related to the rasterization of the individual indexed quads that become 2 triangles of the circle each. It just approximates the mathematical defined borders to pixels and on some cases it choses not to add a pixel.

The solution to this issue is drawing using plain triangles or a triangle-strip but it's not possible to get it with quads...

It's a guess but I don't really understand the issue, probably I don't really understand what MSAAx4 really does internally (beside the theorical definition).

**Comment by raysan5** [comment]
I'm afraid this issue is out-of-scope of raylib. It seems it's more related to the driver implementation of OpenGL.

**Comment by evanrinehart** [comment]
I got the exact effect on my NVIDIA linux box and macbook air. I think a different way of drawing a circle with quads might help.

Currently I'm disabling SUPPORT_QUADS_DRAW_MODE in config.h which is enabled by default. It's a shame the default behavior is bugged.

I might be able to look at an alternative circle-with-quads drawing strategy at some point. I'll have to remember this closed issue.

**Comment by evanrinehart** [comment]
Well after a few minutes of messing around I noticed that disabling the 1x1 pixel texture that goes with each slice of `DrawCircleSector`, the erroneous results seem to go away. I'll have to do more testing.

In the mean time, what is the purpose of the dummy texture used in all the QUAD mode rshapes routines?

**Comment by raysan5** [comment]
> I might be able to look at an alternative circle-with-quads drawing strategy at some point.

I tried this for a couple of hours and I couldn't find a solution.

 > what is the purpose of the dummy texture used in all the QUAD mode rshapes routines?

because raylib uses the same internal shader for shapes drawing and textures drawing, and that shader requires a texture. Also note that by default the font texture is used for shapes drawing, not the 1x1 pixel white texture.

**Comment by raysan5** [comment]
I'm reopening it to ask the X experts about it, maybe someone could provide some more info about the issue.

**Comment by raysan5** [comment]
@mausimus you are right! thank you very much for finding it! Indeed there is a 1px white padding around the texture piece used for drawing but it seems it's not enough for MSAAx4! Reviewing it right now!

**Comment by raysan5** [comment]
Not completed yet, just noticed that as smaller is the circle, the bigger should be the padding... very weird issue...

**Comment by evanrinehart** [comment]
OH! The font texture. Of course. I'm glad you guys figured it out.

**Comment by mausimus** [comment]
You're welcome @raysan5 @evanrinehart ! Yes the problematic scenario is when a triangle doesn't cover the center of a pixel but covers one or more MSAA samples within it - the pixel shader is run only once using center of the pixel (which being outside of shape extrapolates UV to out of bounds) and its color output is used by all MSAA samples.

**Comment by verioneuvien** [comment]
This bug reappears in the DrawRectangleRounded function:

![image](https://github.com/raysan5/raylib/assets/167703509/ef76e8d1-c60d-4c49-abb9-2b424b5dffbb)

Image generated with the below code
```cpp
#include <raylib.h>

int main()
{
	SetConfigFlags(FLAG_MSAA_4X_HINT);
	InitWindow(800, 400, "Rounded Rectangle Pixel Awkward\n");

	int segmentss[4] = { 1, 30, 45, 90 };
	int radiuss[4] = { 5, 10, 20, 40 };

	while (!WindowShouldClose())
	{
		BeginDrawing();
		ClearBackground(WHITE);

		for (int s = 0; s < 4; s++)
		{
			for (int r = 0; r < 4; r++)
			{
				// Calculate position and stuff
				Rectangle rec = { s * 200.0f + 20.0f, r * 100.0f + 10.0f, 160.0f, 80.0f };
				float roundness = radiuss[r] * 2.0f / rec.height; // WHY??

				// Main thing
				DrawRectangleRounded(rec, roundness, segmentss[s], BLACK);

				// Draw text on it
				const char *text = TextFormat("R:%d,S:%d", radiuss[r], segmentss[s]);
				float textWidth = MeasureText(text, 20);
				DrawText(text, rec.x + (rec.width - textWidth) / 2, rec.y + (rec.height - 20) / 2, 20, WHITE);
			}
		}

		EndDrawing();
	}

	CloseWindow();
}
```

System: Arch Linux x86_64
Raylib version: just pulled from master

Raylib logs:

INFO: Initializing raylib 5.1-dev
INFO: Platform backend: DESKTOP (GLFW)
INFO: Supported raylib modules:
INFO:     > rcore:..... loaded (mandatory)
INFO:     > rlgl:...... loaded (mandatory)
INFO:     > rshapes:... loaded (optional)
INFO:     > rtextures:. loaded (optional)
INFO:     > rtext:..... loaded (optional)
INFO:     > rmodels:... loaded (optional)
INFO:     > raudio:.... loaded (optional)
INFO: DISPLAY: Trying to enable MSAA x4
INFO: DISPLAY: Device initialized successfully
INFO:     > Display size: 1920 x 1080
INFO:     > Screen size:  800 x 400
INFO:     > Render size:  800 x 400
INFO:     > Viewport offsets: 0, 0
WARNING: GLFW: Error: 65548 Description: Wayland: The platform does not support setting the window position
INFO: GLAD: OpenGL extensions loaded successfully
INFO: GL: Supported extensions count: 230
INFO: GL: OpenGL device information:
INFO:     > Vendor:   Intel
INFO:     > Renderer: Mesa Intel(R) Xe Graphics (TGL GT2)
INFO:     > Version:  4.6 (Core Profile) Mesa 24.0.5-arch1.1
INFO:     > GLSL:     4.60
INFO: GL: VAO extension detected, VAO functions loaded successfully
INFO: GL: NPOT textures extension detected, full NPOT textures supported
INFO: GL: DXT compressed textures supported
INFO: GL: ETC2/EAC compressed textures supported
INFO: GLFW platform: Wayland
INFO: PLATFORM: DESKTOP (GLFW): Initialized successfully
INFO: TEXTURE: [ID 1] Texture loaded successfully (1x1 | R8G8B8A8 | 1 mipmaps)
INFO: TEXTURE: [ID 1] Default texture loaded successfully
INFO: SHADER: [ID 1] Vertex shader compiled successfully
INFO: SHADER: [ID 2] Fragment shader compiled successfully
INFO: SHADER: [ID 3] Program shader loaded successfully
INFO: SHADER: [ID 3] Default shader loaded successfully
INFO: RLGL: Render batch vertex buffers loaded successfully in RAM (CPU)
INFO: RLGL: Render batch vertex buffers loaded successfully in VRAM (GPU)
INFO: RLGL: Default OpenGL state initialized successfully
INFO: TEXTURE: [ID 2] Texture loaded successfully (128x128 | GRAY_ALPHA | 1 mipmaps)
INFO: FONT: Default font loaded successfully (224 glyphs)
INFO: TEXTURE: [ID 2] Unloaded texture data from VRAM (GPU)
INFO: SHADER: [ID 3] Default shader unloaded successfully
INFO: TEXTURE: [ID 1] Default texture unloaded successfully

**Comment by Cotterzz** [comment]
I've been looking at this and I am wondering if a different method of drawing curves might give improved rendering?
It's an odd way around the problem, but moving the lines to the edge instead of the centre might look better:
For illustration:
![screenrec007](https://github.com/raysan5/raylib/assets/10715212/59aa112a-0c55-4f7c-b3fd-c191e8ff1f37)
This is with MSAA turned on, though I'm on windows, no idea if this is the way to repeat the bug. Made it small and black for comparison. I'm using a custom mesh object:
![screenrec001](https://github.com/raysan5/raylib/assets/10715212/c37bebd8-7205-4f4f-871e-0e047d6b2f39)

**Comment by raysan5** [comment]
@Cotterzz This is an interesting approach, I didn't know about it! It could be an alternative but the code complexity required seems quite high compared to current solution.

**Comment by Cotterzz** [comment]
Yes, and it's a very specific use case for 2D only, so I can't really say if it's worth implementing.
--
My code is probably very overcomplicated as it was my first attempt at the algorithm, I'm not great with C, and I should probably have used recursion - so if you did want to use it, I'm sure it can be reduced.
